ci-operator/config/openshift/installer/master: Move to openshift-install

wking opened the pull request:
This is the next-gen installer.  With the new installer, we only need the installer binary and terraform to launch and destroy clusters.  I've removed the smoke tests for now, because I haven't looked into building them without Bazel.  Hopefully we'll get expanded e2e testing *and* other OpenShift projects using our installer soon to keep us honest.

Also missing from the new installer is a way to set `expirationDate`.  But we'll just not leak until we regain the ability to set that, right? ;)

The new config dumps less cruft into the output directory (most of the generated output goes into a temporary directory), so I've adjusted the openshift-install calls to just use an artifacts subdir for their state storage.  A bonus of this approach is that if the installer hangs up, we'll capture anything it wrote to disk without needing an explicit `cp` call.

I've also dropped the 3.11 config, since the installer is 4.0-only.

We'll need openshift/installer#333 to land before this PR will work.

/hold

And even once that PR lands, I dunno if this one will work.  @sallyom, can you run this through its paces or give me some pointers?

CC @crawford; once this lands, we can delete the old installer without breaking CI.

crawford commented:
/approve

wking commented:
Rebased around #1690.

smarterclayton commented on ci-operator/templates/cluster-launch-installer-e2e.yaml:
```diff
@@ -195,26 +194,19 @@ objects:
-        cp terraform.tfvars /tmp/artifacts/installer/
+        echo "Invoking installer ..."
+        OPENSHIFT_INSTALL_CLUSTER_NAME="${INSTANCE_PREFIX}" \
+        OPENSHIFT_INSTALL_BASE_DOMAIN=origin-ci-int-aws.dev.rhcloud.com \
```
Be aware that at some point in the next month or so we'll likely need to move this back out of the job definition and make parametrizable for different teams and content streams.  But it's fine for now.

smarterclayton commented on ci-operator/templates/cluster-launch-installer-e2e.yaml:
```diff
@@ -181,7 +180,7 @@ objects:
         value: /etc/openshift-installer/.awscred
       - name: KUBECONFIG
-        value: /tmp/cluster/generated/auth/kubeconfig
+        value: /tmp/artifacts/installer/auth/kubeconfig
```
Putting kubeconfig in artifacts isn't a great idea if we accidentally put a secret in there.

smarterclayton replied:
(i.e. if something shows up in there that is longer lived than the cluster)

wking replied:
> Putting kubeconfig in artifacts isn't a great idea if we accidentally put a secret in there.

I'd guessed that we were we generating the secret for this CI run on the fly.  Is that not accurate?  Or is this just a "watch out for long-lived secrets in the future"?

smarterclayton commented on ci-operator/config/openshift/installer/master.yaml:
```diff
@@ -1,44 +1,21 @@
-  from: installer-bazel
-  to: installer-smoke
+        source_path: /go/src/github.com/openshift/installer/bin/terraform
+  to: installer
```
Is the replacement dockerfile centos based and using the right tooling?

wking replied:
> Is the replacement dockerfile centos based and using the right tooling?

There is no replacement `Dockerfile`.  openshift/release:golang-1.10 has what we need for `hack/build.sh && hack/get-terraform.sh`, and I was hoping to build the `:installer` image with just that [here][1].  But I don't know what I'm doing with ci-operator; maybe that's not how it works?

[1]: https://github.com/openshift/release/blob/37c66462b2d5671f2a44c2067e6e1a7270cef907/ci-operator/config/openshift/installer/master.yaml#L1-L18

wking replied:
> There is no replacement `Dockerfile`.

This was a dead end.  I've pushed a replacement `Dockerfile` with openshift/installer#343, but it's based on `scratch`, not CentOS.  Do we need CentOS for something?

dgoodwin replied:
ART team recommends this, may be outright required, I'm not sure. (https://mojo.redhat.com/docs/DOC-1179058#jive_content_id_Images)

I also quickly noticed I can't exec in with /bin/bash, which is awfully handy for debugging. Actually in the way we're probably going to use this in Hive we probably need a more substantial base layer.

wking replied:
> I also quickly noticed I can't exec in with /bin/bash, which is awfully handy for debugging. 

The `Dockerfile` is stuffing the binaries into `scratch` (for folks who just need the binaries), but I couldn't figure out the right phrasing for that here (lucky for you ;), so we're currently [using the base image][1] (defined [here][2]) as the base layer that gets the binaries added.  I chat about this a bit in the 3f2f01c43 commit message if you want more details.

[1]: https://github.com/openshift/release/blob/3f2f01c43b51725b685621a95a3acc2115ddfb47/ci-operator/config/openshift/installer/master.yaml#L10
[2]: https://github.com/openshift/release/blob/3f2f01c43b51725b685621a95a3acc2115ddfb47/ci-operator/config/openshift/installer/master.yaml#L1-L6

dgoodwin replied:
Ah so your published CI output is quite different from your scratch Dockerfile.ci if you built it yourself? As long as the published image is CentOS/RHEL based I think we'll have what we need.

You could also consider multiple Dockerfiles if need be. I can hack my own for now until this is getting publishing.

smarterclayton commented:
(none of my comments are blocking)

/approve

wking commented:
/hold cancel

openshift/install#340 landed, so we should be ok here, although I still haven't gotten this tested locally (@sallyom's been helping me with [her notes][1], but I've been distracted).

[1]: https://github.com/sallyom/installer-e2e

wking commented:
Rebased around #1688 and updated with a @sallyom suggestion to avoid:

```
error: could not resolve inputs: could not determine inputs for step [input:root]: could not resolve base image: imagestreamtags.image.openshift.io "installer-test-base:" is forbidden: User "wking" cannot get imagestreamtags.image.openshift.io in the namespace "stable": no RBAC policy matched
```

in local testing.

wking commented:
Ok, in local testing with [this slightly-tweaked template][1], I'm getting:

```console
$ ci-operator -template templates/cluster-launch-installer-e2e-new.yaml -config ~/src/openshift/release/ci-operator/config/openshift/installer/master.yaml -git-ref=openshift/installer@master -namespace wking-next-gen-installer
2018/09/26 13:57:05 Resolved openshift/installer@master to commit 60991d84b2a266187c3a46f4bb91662cd15afccc
2018/09/26 13:57:05 Resolved source https://github.com/openshift/installer to master@60991d84
2018/09/26 13:57:06 Resolved openshift/release:golang-1.10 to sha256:43ad4740f25e4cbc6de8a964e2ce65ce50fbc24fc4e5e268e0a7370ca3b09bd1
2018/09/26 13:57:07 Resolved openshift/origin-v4.0:base to sha256:8026efab865dcb2d776b47ce05d639c6cb4113225df75c1fbce785bd6dfc88b7
2018/09/26 13:57:07 Using namespace wking-next-gen-installer
2018/09/26 13:57:07 Running [input:root], [input:base], [release-inputs], src, bin, installer, unit, gofmt, govet, golint, [output:stable:installer], [images], cluster-launch-installer-e2e-new, [release:latest]
2018/09/26 13:57:07 Creating namespace wking-next-gen-installer
2018/09/26 13:57:07 Setting a soft TTL of 1h0m0s for the namespace
2018/09/26 13:57:07 Setting a hard TTL of 12h0m0s for the namespace
2018/09/26 13:57:08 warning: Could not mark the namespace to be deleted later because you do not have permission to update the namespace (details: namespaces "wking-next-gen-installer" is forbidden: User "wking" cannot update namespaces in the namespace "wking-next-gen-installer": no RBAC policy matched)
2018/09/26 13:57:08 Setting up pipeline imagestream for the test
2018/09/26 13:57:09 Populating secrets for test
2018/09/26 13:57:09 Tagging openshift/release:golang-1.10 into pipeline:root
2018/09/26 13:57:09 Tagging openshift/origin-v4.0:base into pipeline:base
2018/09/26 13:57:09 Tagged release images from openshift/origin-v4.0:${component}, images will be pullable from registry.svc.ci.openshift.org/wking-next-gen-installer/stable:${component}
2018/09/26 13:57:09 Building src
2018/09/26 13:57:12 Build src already succeeded in 27s
2018/09/26 13:57:12 Executing test golint
2018/09/26 13:57:12 Executing test gofmt
2018/09/26 13:57:12 Building bin
2018/09/26 13:57:12 Executing test unit
2018/09/26 13:57:12 Executing test govet
2018/09/26 13:57:13 Build bin already succeeded in 35s
2018/09/26 13:57:13 Building installer
2018/09/26 13:57:14 Build installer failed, printing logs:
Pulling image "docker-registry.default.svc:5000/wking-next-gen-installer/pipeline@sha256:d913d470e79f89cf1797725d2aa0884fd9d3424fb902f41a697df3e4f1d708d4" ...
Pulling image "docker-registry.default.svc:5000/wking-next-gen-installer/pipeline@sha256:e3eb8f4e82cfffda0568c0853ddbe11c695b877365be67ed1b77393bea68df43" ...
error: open /tmp/build/inputs/Dockerfile: no such file or directory
2018/09/26 13:57:19 Container test in pod gofmt completed successfully
2018/09/26 13:57:19 Pod gofmt succeeded after 4s
2018/09/26 13:57:21 Container test in pod golint completed successfully
2018/09/26 13:57:21 Pod golint succeeded after 5s
2018/09/26 13:57:33 Container test in pod unit completed successfully
2018/09/26 13:57:33 Pod unit succeeded after 19s
2018/09/26 14:00:24 Container test in pod govet completed successfully
2018/09/26 14:00:24 Pod govet succeeded after 3m10s
2018/09/26 14:00:25 Ran for 3m19s
error: could not run steps: could not wait for build: the build installer failed with reason GenericBuildFailed: Generic Build failure - check logs for details.
```

I dunno where the `/tmp/build/inputs/Dockerfile` is coming from, grepping hasn't turned it up for me in either this repo or ci-operator.  And that may not even be the main issue.

[1]: https://github.com/wking/installer-e2e/commit/6c0ad63712d9009ebaf279ddee47dd30c1680184

wking commented:
Probably something in [here][1] is assuming the presence of a `Dockerfile`, when I'm trying to get it to use `base` as its build image, run `hack/build.sh && hack/get-terraform.sh`, and stick the resulting two binaries into the output `installer` container.

[1]: https://github.com/openshift/release/blob/314420553e3bbf7da637db3020bcb8742aa83bb7/ci-operator/config/openshift/installer/master.yaml#L1-L18

wking commented:
After poking around in ci-operator, it looks like I will need a `Dockerfile` somewhere in the installer repo to build this image (openshift/ci-operator#164).

wking commented:
Where are [`IMAGE_INSTALLER` and `IMAGE_TESTS`][1] set?  I'm not having much luck with `grep`.

[1]: https://github.com/openshift/release/blob/dc9df9fe3275bc439de9f0eb7d213d16a12bf8b4/ci-operator/templates/cluster-launch-installer-e2e.yaml#L13-L16

wking commented:
> Where are `IMAGE_INSTALLER` and `IMAGE_TESTS` set?

Ah, docs are [here][1].

[1]: https://github.com/openshift/ci-operator/blob/58d84eba499dcb1a0bb16d65c48ccdc5a7892ba7/TEMPLATES.md#image_component

abhinavdahiya commented:
https://github.com/openshift/ci-operator/blob/master/cmd/ci-operator/main.go#L80

so `installer` image becomes `IMAGE_INSTALLER` and image `tests` become `IMAGE_TESTS` i suppose.

wking commented:
I've pushed 590673a -> 3f2f01c.  Combined with openshift/installer#345, this gets pretty close to passing for me when I test locally based on @sallyom's [notes][1].  I made these changes locally to trim some reliable cruft (faster iteration ;) and tie into my testing namespace/secret:

```diff
diff --git a/ci-operator/config/openshift/installer/master.yaml b/ci-operator/config/openshift/installer/master.yaml
index 9f451bb..f030c07 100644
--- a/ci-operator/config/openshift/installer/master.yaml
+++ b/ci-operator/config/openshift/installer/master.yaml
@@ -44,12 +44,3 @@ tests:
 - as: unit
   commands: go test ./pkg/...
   from: src
-- as: gofmt
-  commands: IS_CONTAINER=TRUE ./hack/go-fmt.sh .
-  from: src
-- as: govet
-  commands: IS_CONTAINER=TRUE ./hack/go-vet.sh ./...
-  from: src
-- as: golint
-  commands: IS_CONTAINER=TRUE ./hack/go-lint.sh -min_confidence 0.3 $(go list -f '{{ .ImportPath }}' ./...)
-  from: src
diff --git a/ci-operator/templates/cluster-launch-installer-e2e.yaml b/ci-operator/templates/cluster-launch-installer-e2e.yaml
index 898b69f..8b67d1a 100644
--- a/ci-operator/templates/cluster-launch-installer-e2e.yaml
+++ b/ci-operator/templates/cluster-launch-installer-e2e.yaml
@@ -8,6 +8,7 @@ parameters:
   required: true
 - name: NAMESPACE
   required: true
+  value: "wking-next-gen-installer"
 - name: IMAGE_FORMAT
   required: true
 - name: LOCAL_IMAGE_INSTALLER
@@ -16,6 +17,7 @@ parameters:
   required: true
 - name: CLUSTER_TYPE
   required: true
+  value: "aws"
 - name: TEST_FOCUS
 - name: TEST_SKIP
 - name: TEST_FOCUS_SERIAL
@@ -60,7 +62,7 @@ objects:
       emptyDir: {}
     - name: cluster-profile
       secret:
-        secretName: ${JOB_NAME_SAFE}-cluster-profile
+        secretName: cluster-profile
 
     containers:
```

Then testing with:

```console
$ oc delete is pipeline stable -n wking-next-gen-installer  # cleanup from previous runs
imagestream.image.openshift.io "pipeline" deleted
imagestream.image.openshift.io "stable" deleted
$ ci-operator -template ci-operator/templates/cluster-launch-installer-e2e.yaml -config ci-operator/config/openshift/installer/master.yaml -git-ref=wking/openshift-installer@next-gen-dockerfile -namespace wking-next-gen-installer
2018/09/27 03:56:55 Resolved wking/openshift-installer@next-gen-dockerfile to commit 7a1c003be9d124e44c79c86b1f6d1e6605a4a7c8
2018/09/27 03:56:55 Resolved source https://github.com/wking/openshift-installer to next-gen-dockerfile@7a1c003b
2018/09/27 03:56:55 Resolved openshift/release:golang-1.10 to sha256:43ad4740f25e4cbc6de8a964e2ce65ce50fbc24fc4e5e268e0a7370ca3b09bd1
2018/09/27 03:56:55 Resolved openshift/origin-v4.0:base to sha256:25872551dc91b808915d8f880635eb181ea5520d43c0fd53f77bdf4a84380a8f
2018/09/27 03:56:55 Using namespace wking-next-gen-installer
2018/09/27 03:56:55 Running [input:root], [input:base], [release-inputs], src, installer, [output:stable:installer], [images], unit, cluster-launch-installer-e2e, [release:latest]
2018/09/27 03:56:55 Creating namespace wking-next-gen-installer
2018/09/27 03:56:56 Setting a soft TTL of 1h0m0s for the namespace
2018/09/27 03:56:56 Setting a hard TTL of 12h0m0s for the namespace
2018/09/27 03:56:56 warning: Could not mark the namespace to be deleted later because you do not have permission to update the namespace (details: namespaces "wking-next-gen-installer" is forbidden: User "wking" cannot update namespaces in the namespace "wking-next-gen-installer": no RBAC policy matched)
2018/09/27 03:56:56 Setting up pipeline imagestream for the test
2018/09/27 03:56:56 Populating secrets for test
2018/09/27 03:56:56 Tagging openshift/release:golang-1.10 into pipeline:root
2018/09/27 03:56:56 Tagging openshift/origin-v4.0:base into pipeline:base
2018/09/27 03:56:56 Tagged release images from openshift/origin-v4.0:${component}, images will be pullable from registry.svc.ci.openshift.org/wking-next-gen-installer/stable:${component}
2018/09/27 03:56:56 Building src
2018/09/27 03:59:53 Build src succeeded after 2m55s
2018/09/27 03:59:53 Executing test unit
2018/09/27 03:59:53 Building installer
2018/09/27 04:00:22 Container test in pod unit completed successfully
2018/09/27 04:00:22 Pod unit succeeded after 26s
2018/09/27 04:01:23 Build installer succeeded after 1m30s
2018/09/27 04:01:23 Tagging installer into stable
2018/09/27 04:01:23 All images ready
2018/09/27 04:01:23 Executing template cluster-launch-installer-e2e
2018/09/27 04:01:23 Creating or restarting template instance
2018/09/27 04:01:23 Create a new update payload image registry.svc.ci.openshift.org/wking-next-gen-installer/release:latest
2018/09/27 04:01:23 Executing release release-latest
2018/09/27 04:01:24 Waiting for template instance to be ready
2018/09/27 04:01:26 Running pod dev
time="2018-09-27T11:01:28Z" level=info msg="Fetching Terraform Variables..." 
time="2018-09-27T11:01:28Z" level=debug msg="Generating dependencies of Terraform Variables..." 
time="2018-09-27T11:01:28Z" level=info msg="  Fetching Install Config..." 
time="2018-09-27T11:01:28Z" level=debug msg="  Generating dependencies of Install Config..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Cluster ID..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Cluster ID..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Email Address..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Email Address..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Password..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Password..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching SSH Key..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating SSH Key..." 
? SSH Public Key  [Use arrows to move, type to filter, ? for more help]
> <none>
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Base Domain..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Base Domain..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Cluster Name..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Cluster Name..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Pull Secret..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Pull Secret..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Platform..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Platform..." 
time="2018-09-27T11:01:28Z" level=debug msg="  Generating Install Config..." 
time="2018-09-27T11:01:28Z" level=info msg="  Fetching Bootstrap Ignition Config..." 
time="2018-09-27T11:01:28Z" level=debug msg="  Generating dependencies of Bootstrap Ignition Config..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Install Config..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Found Install Config..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Root CA..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Root CA..." 
time="2018-09-27T11:01:28Z" level=info msg="    Fetching Certificate (etcd)..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating dependencies of Certificate (etcd)..." 
time="2018-09-27T11:01:28Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:28Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:28Z" level=debug msg="    Generating Certificate (etcd)..." 
time="2018-09-27T11:01:29Z" level=info msg="    Fetching Certificate ()..." 
time="2018-09-27T11:01:29Z" level=debug msg="    Generating dependencies of Certificate ()..." 
time="2018-09-27T11:01:29Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:29Z" level=debug msg="      Generating dependencies of Certificate (kube-ca)..." 
time="2018-09-27T11:01:29Z" level=info msg="        Fetching Root CA..." 
time="2018-09-27T11:01:29Z" level=debug msg="        Found Root CA..." 
time="2018-09-27T11:01:29Z" level=debug msg="      Generating Certificate (kube-ca)..." 
time="2018-09-27T11:01:29Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:29Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:29Z" level=debug msg="    Generating Certificate ()..." 
time="2018-09-27T11:01:30Z" level=info msg="    Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:30Z" level=debug msg="    Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:30Z" level=info msg="    Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:30Z" level=debug msg="    Generating dependencies of Certificate (aggregator)..." 
time="2018-09-27T11:01:30Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:30Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:30Z" level=debug msg="    Generating Certificate (aggregator)..." 
time="2018-09-27T11:01:30Z" level=info msg="    Fetching Certificate (service-serving)..." 
time="2018-09-27T11:01:30Z" level=debug msg="    Generating dependencies of Certificate (service-serving)..." 
time="2018-09-27T11:01:30Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:30Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:30Z" level=debug msg="    Generating Certificate (service-serving)..." 
time="2018-09-27T11:01:31Z" level=info msg="    Fetching Certificate (clusterapi.openshift-cluster-api.svc)..." 
time="2018-09-27T11:01:31Z" level=debug msg="    Generating dependencies of Certificate (clusterapi.openshift-cluster-api.svc)..." 
time="2018-09-27T11:01:31Z" level=info msg="      Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:31Z" level=debug msg="      Found Certificate (aggregator)..." 
time="2018-09-27T11:01:31Z" level=debug msg="    Generating Certificate (clusterapi.openshift-cluster-api.svc)..." 
time="2018-09-27T11:01:31Z" level=info msg="    Fetching Certificate (etcd)..." 
time="2018-09-27T11:01:31Z" level=debug msg="    Generating dependencies of Certificate (etcd)..." 
time="2018-09-27T11:01:31Z" level=info msg="      Fetching Certificate (etcd)..." 
time="2018-09-27T11:01:31Z" level=debug msg="      Found Certificate (etcd)..." 
time="2018-09-27T11:01:31Z" level=debug msg="    Generating Certificate (etcd)..." 
time="2018-09-27T11:01:32Z" level=info msg="    Fetching Certificate (kube-apiserver)..." 
time="2018-09-27T11:01:32Z" level=debug msg="    Generating dependencies of Certificate (kube-apiserver)..." 
time="2018-09-27T11:01:32Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:32Z" level=debug msg="      Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:32Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:32Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:32Z" level=debug msg="    Generating Certificate (kube-apiserver)..." 
time="2018-09-27T11:01:33Z" level=info msg="    Fetching Certificate (openshift-apiserver)..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating dependencies of Certificate (openshift-apiserver)..." 
time="2018-09-27T11:01:33Z" level=info msg="      Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:33Z" level=debug msg="      Found Certificate (aggregator)..." 
time="2018-09-27T11:01:33Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:33Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating Certificate (openshift-apiserver)..." 
time="2018-09-27T11:01:33Z" level=info msg="    Fetching Certificate (kube-apiserver-proxy)..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating dependencies of Certificate (kube-apiserver-proxy)..." 
time="2018-09-27T11:01:33Z" level=info msg="      Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:33Z" level=debug msg="      Found Certificate (aggregator)..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating Certificate (kube-apiserver-proxy)..." 
time="2018-09-27T11:01:33Z" level=info msg="    Fetching Certificate (system:admin)..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating dependencies of Certificate (system:admin)..." 
time="2018-09-27T11:01:33Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:33Z" level=debug msg="      Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:33Z" level=debug msg="    Generating Certificate (system:admin)..." 
time="2018-09-27T11:01:34Z" level=info msg="    Fetching Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:34Z" level=debug msg="    Generating dependencies of Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:34Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:34Z" level=debug msg="      Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:34Z" level=debug msg="    Generating Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:34Z" level=info msg="    Fetching Certificate ()..." 
time="2018-09-27T11:01:34Z" level=debug msg="    Generating dependencies of Certificate ()..." 
time="2018-09-27T11:01:34Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:34Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:34Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:34Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:34Z" level=debug msg="    Generating Certificate ()..." 
time="2018-09-27T11:01:34Z" level=info msg="    Fetching Key Pair (service-account.pub)..." 
time="2018-09-27T11:01:34Z" level=debug msg="    Generating Key Pair (service-account.pub)..." 
time="2018-09-27T11:01:35Z" level=info msg="    Fetching Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=debug msg="    Generating dependencies of Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Certificate (system:admin)..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Certificate (system:admin)..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="    Generating Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=info msg="    Fetching Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=debug msg="    Generating dependencies of Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="    Generating Kubeconfig..." 
time="2018-09-27T11:01:35Z" level=info msg="    Fetching Common Manifests..." 
time="2018-09-27T11:01:35Z" level=debug msg="    Generating dependencies of Common Manifests..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Kube Core Operator..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating dependencies of Kube Core Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="        Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="        Found Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating Kube Core Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Network Operator..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating dependencies of Network Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="        Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="        Found Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating Network Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Kube Addon Operator..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating dependencies of Kube Addon Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="        Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="        Found Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating Kube Addon Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="      Fetching Machine API Operator..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating dependencies of Machine API Operator..." 
time="2018-09-27T11:01:35Z" level=info msg="        Fetching Install Config..." 
time="2018-09-27T11:01:35Z" level=debug msg="        Found Install Config..." 
time="2018-09-27T11:01:35Z" level=info msg="        Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:35Z" level=debug msg="        Found Certificate (aggregator)..." 
time="2018-09-27T11:01:35Z" level=debug msg="      Generating Machine API Operator..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Generating dependencies of Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=info msg="        Fetching Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="        Found Install Config..." 
time="2018-09-27T11:01:36Z" level=info msg="        Fetching Certificate ()..." 
time="2018-09-27T11:01:36Z" level=debug msg="        Found Certificate ()..." 
time="2018-09-27T11:01:36Z" level=info msg="        Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=debug msg="        Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Generating Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Root CA..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Root CA..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (etcd)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (etcd)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate ()..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate ()..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (aggregator)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (aggregator)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (service-serving)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (service-serving)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (clusterapi.openshift-cluster-api.svc)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (clusterapi.openshift-cluster-api.svc)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (etcd)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (etcd)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (kube-apiserver)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (kube-apiserver)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (openshift-apiserver)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (openshift-apiserver)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (kube-apiserver-proxy)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (kube-apiserver-proxy)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate ()..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate ()..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (system:serviceaccount:kube-system:default)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Key Pair (service-account.pub)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Key Pair (service-account.pub)..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Kubeconfig..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Kubeconfig..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Generating Common Manifests..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Generating dependencies of Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate ()..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate ()..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Certificate (kube-ca)..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Generating Tectonic Manifests..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Kube Core Operator..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Generating dependencies of Kube Core Operator..." 
time="2018-09-27T11:01:36Z" level=info msg="      Fetching Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="      Found Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Generating Kube Core Operator..." 
time="2018-09-27T11:01:36Z" level=debug msg="  Generating Bootstrap Ignition Config..." 
time="2018-09-27T11:01:36Z" level=info msg="  Fetching Master Ignition Config(s)..." 
time="2018-09-27T11:01:36Z" level=debug msg="  Generating dependencies of Master Ignition Config(s)..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Found Install Config..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Root CA..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Found Root CA..." 
time="2018-09-27T11:01:36Z" level=debug msg="  Generating Master Ignition Config(s)..." 
time="2018-09-27T11:01:36Z" level=info msg="  Fetching Worker Ignition Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="  Generating dependencies of Worker Ignition Config..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Install Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Found Install Config..." 
time="2018-09-27T11:01:36Z" level=info msg="    Fetching Root CA..." 
time="2018-09-27T11:01:36Z" level=debug msg="    Found Root CA..." 
time="2018-09-27T11:01:36Z" level=debug msg="  Generating Worker Ignition Config..." 
time="2018-09-27T11:01:36Z" level=debug msg="Generating Terraform Variables..." 
time="2018-09-27T11:01:36Z" level=fatal msg="failed to generate asset: failed to determine default AMI: MissingRegion: could not find region configuration" 
? Pull Secret [? for help]                                                                                                                                                                                        2018/09/27 04:01:37 Container setup in pod dev failed, exit code 1, reason Error
^C
2018/09/27 04:02:09 error: Process interrupted with signal interrupt, exiting in 2s ...
2018/09/27 04:02:09 cleanup: Deleting release pod release-latest
2018/09/27 04:02:09 cleanup: Deleting template cluster-launch-installer-e2e
2018/09/27 04:02:09 cleanup: Deleting test pod unit
```

I dunno why it's complaining about a missing region (it doesn't seem to have asked for a region), but that feels like an installer bug, and not a bug with this PR.  I'm certainly [trying to set `OPENSHIFT_INSTALL_AWS_REGION`][2].

[1]: https://github.com/sallyom/installer-e2e
[2]: https://github.com/openshift/release/commit/3f2f01c43b51725b685621a95a3acc2115ddfb47#diff-4b206e1e0e8bf659dda6283fbd812433R226

dgoodwin commented:
Regarding missing region see https://github.com/openshift/installer/issues/335. Fairly confident that the Terraform is not respecting the same methods and ordering of loading AWS credentials as the installer.

wking commented:
> Regarding missing region see openshift/installer#335.

Ah, thanks for the reproduction instructions :).  Does openshift/installer#346 fix it for you?

wking commented:
A bunch more installer work later, and now I'm running:

```diff
diff --git a/ci-operator/config/openshift/installer/master.yaml b/ci-operator/config/openshift/installer/master.yaml
index 9f451bb..f030c07 100644
--- a/ci-operator/config/openshift/installer/master.yaml
+++ b/ci-operator/config/openshift/installer/master.yaml
@@ -44,12 +44,3 @@ tests:
 - as: unit
   commands: go test ./pkg/...
   from: src
-- as: gofmt
-  commands: IS_CONTAINER=TRUE ./hack/go-fmt.sh .
-  from: src
-- as: govet
-  commands: IS_CONTAINER=TRUE ./hack/go-vet.sh ./...
-  from: src
-- as: golint
-  commands: IS_CONTAINER=TRUE ./hack/go-lint.sh -min_confidence 0.3 $(go list -f '{{ .ImportPath }}' ./...)
-  from: src
diff --git a/ci-operator/templates/cluster-launch-installer-e2e.yaml b/ci-operator/templates/cluster-launch-installer-e2e.yaml
index 898b69f..23b2c88 100644
--- a/ci-operator/templates/cluster-launch-installer-e2e.yaml
+++ b/ci-operator/templates/cluster-launch-installer-e2e.yaml
@@ -8,6 +8,7 @@ parameters:
   required: true
 - name: NAMESPACE
   required: true
+  value: "wking-next-gen-installer"
 - name: IMAGE_FORMAT
   required: true
 - name: LOCAL_IMAGE_INSTALLER
@@ -16,6 +17,7 @@ parameters:
   required: true
 - name: CLUSTER_TYPE
   required: true
+  value: "aws"
 - name: TEST_FOCUS
 - name: TEST_SKIP
 - name: TEST_FOCUS_SERIAL
@@ -60,7 +62,7 @@ objects:
       emptyDir: {}
     - name: cluster-profile
       secret:
-        secretName: ${JOB_NAME_SAFE}-cluster-profile
+        secretName: cluster-profile
 
     containers:
 
@@ -212,7 +214,7 @@ objects:
       - name: OPENSHIFT_INSTALL_CLUSTER_NAME
         value: ${NAMESPACE}-${JOB_NAME_HASH}
       - name: OPENSHIFT_INSTALL_BASE_DOMAIN
-        value: origin-ci-int-aws.dev.rhcloud.com
+        value: coreservices.team.coreos.systems
       - name: OPENSHIFT_INSTALL_EMAIL_ADDRESS
         value: test@ci.openshift.io
       - name: OPENSHIFT_INSTALL_PASSWORD
```

(with the base domain change because I'm using teamcoreservices credentials to launch the cluster).  Launched with:

```console
$ oc delete is pipeline stable -n wking
$ ci-operator -template ci-operator/templates/cluster-launch-installer-e2e.yaml -config ci-operator/config/openshift/installer/master.yaml -git-ref=openshift/installer@master -namespace wking
```

the setup container runs to a successful completion:

```
Apply complete! Resources: 154 added, 0 changed, 0 destroyed.
...
time="2018-09-27T22:23:14Z" level=debug msg="Generating Cluster Metadata..." 
```

The test container is still waiting:

```
Waiting for API at https://wking-ef260-api.coreservices.team.coreos.systems:6443 to respond ... 
```

but the API returns something when I hit it directly:

```console
$ curl -ks https://wking-ef260-api.coreservices.team.coreos.systems:6443/ | head
{
  "paths": [
    "/api",
    "/api/v1",
    "/apis",
    "/apis/",
    "/apis/admissionregistration.k8s.io",
    "/apis/admissionregistration.k8s.io/v1beta1",
    "/apis/apiextensions.k8s.io",
    "/apis/apiextensions.k8s.io/v1beta1",
```

So it's possible that I just need to be patient.  Or it's possible that I broke the wait-for-the-API logic.

wking commented:
Anyhow, I think this is close enough that I'd like to land it, and if there are any remaining bugs we can sort them out on the fly.  @smarterclayton, can you take another look and drop a `/lgtm` if you like it?

wking commented:
The cluster continues:

```
API at https://wking-ef260-api.coreservices.team.coreos.systems:6443  has responded
Waiting for router to be created ...
```

so better and better :).

wking commented:
I ctrl-ced `ci-operator` (because I wanted to fix my patched namespace from `wking-next-gen-installer` -> `wking`, and checking the logs for the teardown cluster, things look healthy there too:

```
...
time="2018-09-27T22:39:40Z" level=debug msg="Deleting subnets"
time="2018-09-27T22:39:40Z" level=debug msg="Exiting deleting subnets"
time="2018-09-27T22:39:40Z" level=debug msg="goroutine deleteSubnets complete"
time="2018-09-27T22:39:42Z" level=debug msg="Deleting security groups"
time="2018-09-27T22:39:42Z" level=debug msg="Exiting deleting security groups"
time="2018-09-27T22:39:42Z" level=debug msg="goroutine deleteSecurityGroups complete"
...
```

crawford commented:
/lgtm
